lib/IDE/CodeCompletion.cpp

```diff
@@ -4637,6 +4638,18 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
                              CurrDeclContext, IncludeInstanceMembers,
                              /*includeDerivedRequirements*/false,
                              /*includeProtocolExtensionMembers*/true);
+    // We not only allow referencing nested types/typealiases directly, but also
+    // qualified by the current type. Thus also suggest current self type so the
+    // user can do a memberwise lookup on it.
+    if (auto SelfType = typeContext->getSelfNominalTypeDecl()) {
+      addNominalTypeRef(SelfType, DeclVisibilityKind::LocalVariable,
+                        DynamicLookupInfo());
+    }
```

Comment:
This doesn't seem to be true when the type is a `protocol` 
This is invalid.
```swift
protocol P {
  associated type T where P.T: Equatable
}
```
whereas this is valid
```swift
protocol P {
  associated type T where T: Equatable
}
```

Reply:
But this kind of code is valid, so I guess we should be suggesting `P`:

```swift
protocol Foo {}
protocol P {
  typealias U = Foo
  associatedtype T where P.U == T
}
```

Reply:
Also, we probably shouldn't be suggesting `Self` in non-final classes, because it resolves to `DynamicSelfType` and triggers an error, but I guess it's more a matter of whether we want to be predictive here (we choose to be restrictive, but almost anything is valid in the event of a same-type constraint).

Reply:
We don’t. See https://github.com/apple/swift/pull/36607/files#diff-e5a9013b7193640e854cee0b254e7d56587de6940c318c2d2121b258d61101beR81